library/nsxt_upgrade_run.py

```diff
@@ -0,0 +1,270 @@
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: nsxt_upgrade_run
+short_description: 'Start the upgrade'
+description: 'Upgrade will start as per the upgrade plan.'
+version_added: '2.7'
+author: 'Rahul Raghuvanshi'
+options:
+    hostname:
+        description: 'Deployed NSX manager hostname.'
+        required: true
+        type: str
+    username:
+        description: 'The username to authenticate with the NSX manager.'
+        required: true
+        type: str
+    password:
+        description: 'The password to authenticate with the NSX manager.'
+        required: true
+        type: str
+    paused_upgrade:
+        description: 'Mode of upgrade'
+        required: true
+        type: bool
+'''
+
+EXAMPLES = '''
+- name: Runs the upgrade
+  nsxt_upgrade_run:
+      hostname: "10.192.167.137"
+      username: "admin"
+      password: "Admin!23Admin"
+      validate_certs: False
+      paused_upgrade: True
+'''
+
+RETURN = '''# '''
+
+import json, time
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.vmware_nsxt import vmware_argument_spec, request
+from ansible.module_utils.common_utils import get_attribute_from_endpoint, clean_and_get_params, get_upgrade_orchestrator_node
+from ansible.module_utils._text import to_native
+
+def get_upgrade_status(module, manager_url, mgr_username, mgr_password, validate_certs):
+  '''
+  Get the current status of upgrade at the start.
+  Doesn't upgrade if any component is in progress 
+  or system is already upgraded.
+  '''
+  upgrade_status = get_attribute_from_endpoint(module, manager_url, '/upgrade/summary',
+                        mgr_username, mgr_password, validate_certs, 'upgrade_status', 
+                        False)
+
+  if upgrade_status == 'IN_PROGRESS' or upgrade_status == 'PAUSING':
+    module.fail_json(msg='Upgrade is in state: %s, can\'t continue' % upgrade_status)
+
+  if upgrade_status == 'SUCCESS':
+    module.exit_json(changed=False, message='Upgrade state is SUCCESS. No need to'
+                    ' continue.')
+  return upgrade_status
+
+def decide_next_step(module, manager_url, mgr_username, mgr_password, 
+                     validate_certs, can_continue, is_failed):
+  '''
+  params:
+  - can_continue: if upgrade can be continued 
+  - is_failed: Is there any component Failure
+  return:
+  - Decides the next operation to be done based on 
+    can_continue and is_failed values 
+  '''
+  if can_continue and is_failed:
+    return
+  elif can_continue and not is_failed:
+    return
+  elif not can_continue and is_failed:
+    raise Exception('Upgrade failed. Please run upgrade status summary'
+                    ' to see the reason of upgrade failure.')
+  else:
+    time.sleep(15)
```

Comment:
why this wait

Reply:
Sometimes as per my observations, it takes time to get overall_status to stabilize. Thus a wait.